Closes #5503: ISO 8601 date in UI and alternative format as tooltip

ypid opened the pull request:
<!--
    Thank you for your interest in contributing to NetBox! Please note
    that our contribution policy requires that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
    IF YOUR PULL REQUEST DOES NOT REFERENCE AN ACCEPTED BUG REPORT OR
    FEATURE REQUEST, IT WILL BE MARKED AS INVALID AND CLOSED.
-->
### Fixes: #5503
<!--
    Please include a summary of the proposed changes below.
-->

With this PR all dates in the UI are now consistently displayed.

I changed the long date format as suggested by @xkilian and confirmed by my own research.

* DATETIME_FORMAT
  * Before: July 20, 2020 4:52 p.m.
  * Now: 20th July, 2020 16:52

"20th July, 2020" would be spoken as "the 20th of July, 2020" but the "the" and "of" are never written.

I also switch from the illogical a.m./p.m. time format to military time format that can be pronounced as "sixteen hundred and fifty-two".

ypid committed:
Closes #5503: ISO 8601 date in UI and alternative format as tooltip

With this commit all dates in the UI are now consistently displayed.

I changed the long date format as suggested by @xkilian and confirmed by my own
research.

* DATETIME_FORMAT
 * Before July 20, 2020 4:52 p.m.
 * Now 20th July, 2020 16:52

"20th July, 2020" would be spoken as "the 20th of July, 2020" but the "the" and
"of" are never written.

The only exception is `object_list.html`. I tried it but there it does not
work so easily because the dates are passed to Jinja as SafeString.

DanSheps commented:
There seems to be a lot of repetition here, do you think it might be possible to create an include and pass the date, it would also make future modifications easier.

ypid commented:
Thanks for the pointer (am was/am new to Django), done.

The only place (that I know of, and I also greped the source and so on) with dates without tooltip is `object_list.html`. I tried it but there it does not work so easily because the dates are passed to Jinja as SafeString. Also, I am not sure if it would be good UI design to have tooltips in the list.

ypid commented:
I rebased and updated this patch to the latest develop branch. Some new date fields were introduced with NetBox 2.11. I hope I found all of them by greping the source. If not, it is not the end of the world. I will come across them eventually and fix them (potentially in a new PR). I applied this patch (including the dependency patch 2618dde1e2837fd4c587bed89d2358199176f01e) on my instance now for testing during regular use. I tested all modified pages manually already. Feel free to have a look.

ypid commented:
I have been using this ever since and did not run into dates that I missed nor other issues. I would say this is ready. @jeremystretch Can you have a look?

jeremystretch commented:
@ypid Adding inclusion tags to display a value (i.e. `{% include 'inc/date_now_span.html' %}`) should be considered poor practice. Instead, we should use a template filter for this (e.g. `{{ mydate|with_current_time }}`). We should also only need the one filter; the `parse_date()` and `datatype_str()` are unnecessary since we can process the value directly.

ypid commented:
I improved it, thanks. I agree with you that this approach is better. Now with the `date_span` filter, it is very easy to just append that if new date files are added.

jeremystretch commented on netbox/netbox/configuration.example.py:
```diff
@@ -261,7 +261,7 @@
 DATE_FORMAT = 'N j, Y'
 SHORT_DATE_FORMAT = 'Y-m-d'
-TIME_FORMAT = 'g:i a'
+TIME_FORMAT = 'H:i:s'
```
Why have you changed the default formats?

ypid replied:
I separated this change into it’s own commit to describe my intend:

> a.m./p.m. is illogical. Military time format can be pronounced as "sixteen hundred and fifty-two".

But that is only the format that appears on hover now. Should I drop that change or the whole commit? Are you fine with `SHORT_DATETIME_FORMAT = 'Y-m-d H:i:s'`?

jeremystretch replied:
Let's just leave the defaults in the example alone. If there's a case to change them, it should be addressed under a separate issue.

ypid replied:
I dropped the commit.

jeremystretch commented on netbox/templates/base.html:
```diff
@@ -67,7 +67,7 @@ <h4><i class="mdi mdi-alert"></i> Maintenance Mode</h4>
                 </div>
                 <div class="col-xs-4 text-center">
-                    <p class="text-muted">{% now 'Y-m-d H:i:s T' %}</p>
+                    <p class="text-muted">{% now_span %} {% now 'T' %}</p>
```
The template tag should be able to render the entire date string.

ypid replied:
The problem with the old code was that the date format was hard coded. I decided to use two template tags to respect the configured date format while preserving the time zone which is not intended to be part of the configurable date format strings. What do you have in mind? Do you want to keep the hard coded format?

jeremystretch commented on netbox/utilities/templatetags/helpers.py:
```diff
@@ -137,6 +139,39 @@ def tzoffset(value):
 
 
+@register.filter(expects_localtime=True)
+def date_span(date_value):
```
I thought I mentioned this previously, but maybe not. Let's avoid the term "span" in the filter name, since it naturally implies a span of time between a start and end time (rather than an HTML `<span>` element). `annotated_date` maybe?

ypid replied:
Makes sense. I appreciate that you care about such things!

jeremystretch commented on netbox/utilities/templatetags/helpers.py:
```diff
@@ -137,6 +139,39 @@ def tzoffset(value):
+    if not date_value:
+        return ''
+
+    if type(date_value) == str:
```
I wouldn't think we'd ever want to apply this tag to a string; we should always be working with datetime objects, no?

ypid replied:
Here the date is a string: https://github.com/netbox-community/netbox/blob/develop/netbox/templates/inc/custom_fields_panel.html

jeremystretch replied:
It'd be better to have the custom field return a date object than to accommodate string values in the template filter. Let's just omit custom field dates for now to keep this from getting any more complex.

ypid replied:
Fine for me, done.

jeremystretch commented on netbox/templates/users/userkey.html:
```diff
@@ -19,7 +20,9 @@ <h4>
-            <small class="text-muted">Created {{ object.created }} &middot; Updated <span title="{{ object.last_updated }}">{{ object.last_updated|timesince }}</span> ago</small>
+            <small class="text-muted">
+              Created {{ object.created|annotated_date }} &middot;
+              Updated <span title="{{ object.last_updated|date:'SHORT_DATETIME_FORMAT' }} ({{ object.last_updated|date:'DATETIME_FORMAT' }})">{{ object.last_updated|timesince }}</span> ago
```
```suggestion
              Updated {{ object.last_updated|annotated_date }} ({{ object.last_updated|timesince }} ago)
```

jeremystretch commented on netbox/templates/generic/object.html:
```diff
@@ -42,8 +42,8 @@
-      Created {{ object.created }} &middot;
-      Updated <span title="{{ object.last_updated }}">{{ object.last_updated|timesince }}</span> ago
+      Created {{ object.created|annotated_date }} &middot;
+      Updated <span title="{{ object.last_updated|date:'SHORT_DATETIME_FORMAT' }} ({{ object.last_updated|date:'DATETIME_FORMAT' }})">{{ object.last_updated|timesince }}</span> ago
```
```suggestion
      Updated {{ object.last_updated|annotated_date }} ({{ object.last_updated|timesince }} ago)
```

jeremystretch commented on netbox/utilities/templatetags/helpers.py:
```diff
@@ -4,8 +4,10 @@
 
 import yaml
 from django import template
+from django.template.defaultfilters import date
```
This looks to be unused?

ypid replied:
It is used by this line for example:

```Python
long_ts = date(date_value, 'DATE_FORMAT')
```

ypid committed:
Do not use annotated_date on custom date fields to avoid date parsing

@jeremystretch:

> It'd be better to have the custom field return a date object than to
> accommodate string values in the template filter. Let's just omit custom
> field dates for now to keep this from getting any more complex.

ypid committed:
Use annotated_date also for updated datetimes

This changes the text from: Updated 5 months, 1 week ago
to: Updated 2021-01-24 00:33 (5 months, 1 week ago)

Co-authored-by: Jeremy Stretch <jstretch@ns1.com>

ypid commented:
Done.
